File: 0121_Best_Time_to_Buy_and_Sell_Stock/main.cpp
// T = O(n), S = O(1)
class Solution {
  public:
    int maxProfit(vector<int> &prices) {
        int n = prices.size();
        if (n < 2) {
            return 0;
        }
        int res = 0, buy = prices[0];
        for (int i = 1; i < n; i++) {
            if (buy > prices[i]) {
                buy = prices[i];
            } else {
                res = max(res, prices[i] - buy);
            }
        }
        return res;
    }
};

File: 0121_Best_Time_to_Buy_and_Sell_Stock/golang/main.go
func maxProfit(prices []int) int {
	if prices == nil || len(prices) == 0 {
		return 0
	}
	n := len(prices)
	buy := prices[0]
	res := 0
	for i := 1; i < n; i++ {
		if prices[i] < buy {
			buy = prices[i]
		} else {
			res = max(res, prices[i]-buy)
		}
	}
	return res
}

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}